src/gtkspellcheck/spellcheck.py

```diff
@@ -495,121 +477,181 @@ def check_range(self, start, end, force_all=False):
 
-        if _pygobject:
-            menu = gtk.Menu.new()
-            group = []
+        action = Gio.SimpleAction.new("ignore-all", GLib.VariantType("s"))
+        action.connect(
+            "activate", lambda _action, word: self.ignore_all(word.get_string())
+        )
+        action_group.add_action(action)
+
+        action = Gio.SimpleAction.new("add-to-dictionary", GLib.VariantType("s"))
+        action.connect(
+            "activate", lambda _action, word: self.add_to_dictionary(word.get_string())
+        )
+        action_group.add_action(action)
+
+        action = Gio.SimpleAction.new("replace-word", GLib.VariantType("s"))
+        action.connect(
+            "activate",
+            lambda _action, suggestion: self._replace_word(suggestion.get_string()),
+        )
+        action_group.add_action(action)
+
+        language = Gio.PropertyAction.new("language", self, "language")
+        action_group.add_action(language)
+
+        self._view.insert_action_group("spelling", action_group)
+
+    def _get_languages_menu(self):
+        if Gtk.MAJOR_VERSION > 3:
+            if self._languages_menu is None:
+                self._languages_menu = self._build_languages_menu()
+            return self._languages_menu
+        else:
+            return self._build_languages_menu()
+
+    def _build_languages_menu(self):
+        if Gtk.MAJOR_VERSION > 3:
+            menu = Gio.Menu.new()
         else:
-            menu = gtk.Menu()
-            group = gtk.RadioMenuItem()
-        connect = []
+
+            def _set_language(item, code):
+                self.language = code
+
+            menu = Gtk.Menu.new()
+            group = []
+            connect = []
+
         for code, name in self.languages:
-            if _pygobject:
-                item = gtk.RadioMenuItem.new_with_label(group, name)
-                group.append(item)
+            if Gtk.MAJOR_VERSION > 3:
+                item = Gio.MenuItem.new(name, None)
+                item.set_action_and_target_value(
+                    "spelling.language", GLib.Variant.new_string(code)
+                )
+                menu.append_item(item)
             else:
-                item = gtk.RadioMenuItem(group, name)
-            if code == self.language:
-                item.set_active(True)
-            connect.append((item, code))
-            menu.append(item)
-        for item, code in connect:
-            item.connect("activate", _set_language, code)
+                item = Gtk.RadioMenuItem.new_with_label(group, name)
+                group.append(item)
+                if code == self.language:
+                    item.set_active(True)
+                connect.append((item, code))
+                menu.append(item)
+        if Gtk.MAJOR_VERSION < 4:
+            for item, code in connect:
+                item.connect("activate", _set_language, code)
+
+        if Gtk.MAJOR_VERSION > 3:
+            return Gio.MenuItem.new_submenu(_("Languages"), menu)
```

Comment:
I think it would be cleaner as follows:

```python
if _IS_GTK3:
    ...
else:
    ...
```

In general, whenever something is done conditionally for either of the versions, there should be an `else` case with the equivalent for the other version or a comment explaining why nothing needs to be done for the other version. That way it's always clear that and how the different versions are handled.

Reply:
Done